lib/dune_file.ml

```diff
@@ -67,6 +67,143 @@ module Lang = struct
   let duniverse_minimum_version = (1, 11)
 end
 
+module Packages = struct
+  open Sexplib0.Sexp
+  module Set = Set.Make (String)
+  module Map = Map.Make (String)
+
+  type t = Random.State.t
+
+  let init = Random.State.make_self_init
+
+  let random_valid_identifier t =
+    let int_range ~start ~end_ = Random.State.int t (end_ - start) + start in
```

Comment:
What's the benefit of using a random identifier over an incrementing counter?

Reply:
Indeed the counter would be simpler, but also more likely to lead to duplicated identical names, given the way the renaming is implemented:

Currently the renaming is done per directory, which is also where the random state is initialized, so each vendored, unpacked tarball gets a new state. If I changed it to use a counter, two tarballs both having an identifier called `bar` that needs renaming and using `%s.a%d` as renaming scheme would end up being called `bar.a0`, thus creating a conflict.

Of course the code could be changed to initialize the deduplication counter once per deduplication run and then have all tarballs use the same counter but that I feel keeping it more separated makes it slightly nicer, since it shares less state overall and instead relies more on the randomness of 8-ASCII characters.

If you have strong opinions on that or good arguments for using a counter that I have missed, I'll gladly revise this code.

Reply:
I'm not sure I follow why there's a need for a random ID, but I think it can be an issue to add some nondeterminism in the `lock` step. Is there a way to use a deterministic function for this, like a hash of package names in collision or something like that?

Reply:
The non-determinism is in the pulling not the locking but indeed you're right, being able to derive the distinction value from a hash of the folder names (which have to be unique already) works better and allows for a deterministic renaming strategy.

I've applied a change to use `Digest.string` and it made the code a bit simpler even.